### ISSUE TITLE
Balancer is not fairly distributing tablets in a heterogeneous cluster during decommission

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylladb#21783]
The [test](https://argus.scylladb.com/tests/scylla-cluster-tests/09f6fd97-b4cd-43c8-a9ab-7a4d41cd6743) that reproduced the problem had 3 nodes with 7 cpus each, later added another 3 smaller nodes with 2 cpu each.

Data set grew to 256 tablets, so the former 3 nodes initially had 256 nodes each due to RF=3. 

The balancer intends to support heterogenous cluster by distributing tablets across nodes based on their average load (tablets / cpus). So a node with fewer cpus will receive less tablets.

When the test decommissioned one small node (depicted by light blue in graph below), it can be noticed that balancer was targeting mostly the other remaining smaller nodes (depicted by green and red), increasing their average load way above the bigger nodes. The graph below shows average load per node.

![Image](https://github.com/user-attachments/assets/c0d8a2c5-375d-416e-9aa5-a8254ad35497)

Logs show that balancer knew the best target, yet it was picking the one with highest average load.
```
Node 694f1312-2244-4483-b5d6-628478535942: rack=1a avg_load=15.5 tablets=31 shards=2 state=being_decommissioned 
Node 1f266a35-cead-4fd1-b004-6b94203d2a3b: rack=1a avg_load=34.5 tablets=69 shards=2 state=normal 
Node 39075f58-3c25-40b8-8f2f-e506c18a5288: rack=1a avg_load=28.571428571428573 tablets=200 shards=7 state=normal 
Node 7321fb27-c50d-4d02-8045-48ace4102972: rack=1a avg_load=35 tablets=70 shards=2 state=normal
Node be4fd614-ffd8-48c8-972f-c94f1956b2f9: rack=1a avg_load=28.428571428571427 tablets=199 shards=7 state=normal
Node e2151317-d43c-4e67-aeca-2bdce8dd5e27: rack=1a avg_load=28.428571428571427 tablets=199 shards=7 state=normal 
...
target node: be4fd614-ffd8-48c8-972f-c94f1956b2f9, avg_load: 28.428571428571427, max: 35
...
Initiating tablet streaming (migration) of 0a8843a0-aff7-11ef-b4c2-b2f813e8b416:52 to 7321fb27-c50d-4d02-8045-48ace4102972:1
```

My theory was that table-aware balancing changes is not working well with heterogeneous clusters. It tries to balance tablets of a given table across nodes, without taking into account the capacity of each individual node. I suspect the problem lies when calculating badness of potential targets. We should review evaluate_dst_badness(), it may have to be adjusted to account for possible heterogeneity of nodes, so smaller nodes will receive less tablets of a given table in comparison to bigger ones (proportionally).

I ran a test that demonstrates that the difference in avg load between nodes remains small with table aware balancing disabled. In this test, there were 2 big (green and yellow) and 2 small nodes (blue and orange). One small node was decommissioned, the remaining small one is depicted by blue. With table aware enabled, we can see the discrepancy in the avg load.

![Image](https://github.com/user-attachments/assets/d5e8d51a-803a-455f-890d-d58a13fc2ec5)

**Comment by raphaelsc** [comment]
yes, node with higher average load is producing lower badness (a metric used to evaluate targets and their impact on balance if they're picked; lower is better)  than one with lower avg load:


```
Node 87193c27-0e5c-4b6a-8001-d170115b3934: rack=rack1 avg_load=10 tablets=20 shards=2 state=being_decommissioned stream_read=0 stream_write=0
Node 512bab43-cc68-4b0b-9c50-35225ca5d8f4: rack=rack1 avg_load=21.5 tablets=215 shards=10 state=normal stream_read=0 stream_write=0
Node d3bb3eb3-2f82-495a-b326-f51d434436a6: rack=rack1 avg_load=31.5 tablets=63 shards=2 state=normal stream_read=0 stream_write=0
Node f09d8a41-5ba5-4310-941a-805cd4e4db0c: rack=rack1 avg_load=21.4 tablets=214 shards=10 state=normal stream_read=0 stream_write=0

target node: f09d8a41-5ba5-4310-941a-805cd4e4db0c, avg_load: 21.4, max: 31.5

Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @512bab43-cc68-4b0b-9c50-35225ca5d8f4:0 node balance threshold: 171, dst: 216 (0.0879)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @d3bb3eb3-2f82-495a-b326-f51d434436a6:0 node balance threshold: 171, dst: 64 (-0.2090)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @f09d8a41-5ba5-4310-941a-805cd4e4db0c:0 node balance threshold: 171, dst: 215 (0.0859)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @d3bb3eb3-2f82-495a-b326-f51d434436a6:0 node balance threshold: 171, dst: 64 (-0.2090)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @512bab43-cc68-4b0b-9c50-35225ca5d8f4:0 node balance threshold: 171, dst: 216 (0.0879)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @d3bb3eb3-2f82-495a-b326-f51d434436a6:0 node balance threshold: 171, dst: 64 (-0.2090)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @f09d8a41-5ba5-4310-941a-805cd4e4db0c:0 node balance threshold: 171, dst: 215 (0.0859)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @d3bb3eb3-2f82-495a-b326-f51d434436a6:0 node balance threshold: 171, dst: 64 (-0.2090)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @512bab43-cc68-4b0b-9c50-35225ca5d8f4:0 node balance threshold: 171, dst: 216 (0.0879)
Table 60fd8cd0-b336-11ef-93a2-f3c94e0fdc19 @d3bb3eb3-2f82-495a-b326-f51d434436a6:0 node balance threshold: 171, dst: 65 (-0.2070)
```
The expectation is that targets @ d3bb3eb3-2f82-495a-b326-f51d434436a6 would have higher badness value (the value in parenthesis) than other nodes with lower avg load, corroborating with the theory that the target evaluation isn't accounting for possible heterogeneity of nodes

**Comment by raphaelsc** [comment]
I think we need to patch evaluate_dst_badness() to think of perfect distribution per node in average load terms, rather than raw load (tablet count) since the current calculation treats all targets as having the same capacity. @tgrabiec do you agree?

**Comment by bhalevy** [comment]
The load by node is less interesting.
We need to look at load per shard, and in the future load / shard / GB when we support heterogeneous shards that manage different amount of storage.

**Comment by raphaelsc** [comment]
> The load by node is less interesting. We need to look at load per shard, and in the future load / shard / GB when we support heterogeneous shards that manage different amount of storage.

This is part of table aware load balancing, where we assign "badness" value for node and shard pair of a given target, and use these values to decide what's the best target. heterogeneous shards?

**Comment by bhalevy** [comment]
> > The load by node is less interesting. We need to look at load per shard, and in the future load / shard / GB when we support heterogeneous shards that manage different amount of storage.
> 
> This is part of table aware load balancing, where we assign "badness" value for node and shard pair of a given target, and use these values to decide what's the best target. heterogeneous shards?

Re: heterogeneous shards: currently the model is that the cluster may have nodes with different shard counts, but the assumption is that the shards are symmetrical, or homogeneous, i.e. they have similar cpus, same amount of memory, serve similar amount of storage (total per node / nr_shards), and have similar network bandwidth (total node bandwidth / nr_shards)

**Comment by raphaelsc** [comment]
> > > The load by node is less interesting. We need to look at load per shard, and in the future load / shard / GB when we support heterogeneous shards that manage different amount of storage.
> > 
> > 
> > This is part of table aware load balancing, where we assign "badness" value for node and shard pair of a given target, and use these values to decide what's the best target. heterogeneous shards?
> 
> Re: heterogeneous shards: currently the model is that the cluster may have nodes with different shard counts, but the assumption is that the shards are symmetrical, or homogeneous, i.e. they have similar cpus, same amount of memory, serve similar amount of storage (total per node / nr_shards), and have similar network bandwidth (total node bandwidth / nr_shards)

ah yes, that's the current assumption, which is why we average the load in each node by cpus, to balance tablets fairly among heterogenous nodes (but homogenous shards, we require today disk:memory:cpu ratio of a given shard to be similar across the cluster). I was wondering why we'd ever need assymetrical shards, where we assign different amount of cpu / memory / disk to shards in a given node.

**Comment by tgrabiec** [comment]
> I think we need to patch evaluate_dst_badness() to think of perfect distribution per node in average load terms, rather than raw load (tablet count) since the current calculation treats all targets as having the same capacity. [@tgrabiec](https://github.com/tgrabiec) do you agree?

Yes. I'll send a patch.
